Add instructions to mock extension methods for resource manager SDKs (#36962)
dotnet/docs: docs/azure/sdk/unit-testing-mocking.md

This approach is useful when you would like to consolidate the dependency creation and share the client between multiple consuming classes.

> ## Understand Azure Resource Manager clients

Reviewer: ```suggestion
## Understand Azure Resource Manager (ARM) clients
```
Maybe unnecessary but ARM seems to be the commonly used term and I'm not confident that customers know the long form of the acronym.

Reviewer 2: I also think we should use the ARM acronym here

> In Azure Resource Manager libraries, the clients were designed to emphasize their relationship to one another, mirroring the service hierarchy. To achieve that goal, extension methods are widely used to add additional features to clients.

Reviewer 2: ```suggestion
In ARM libraries, the clients were designed to emphasize their relationship to one another, mirroring the service hierarchy. To achieve that goal, extension methods are widely used to add additional features to clients.
```

> For example, an Azure virtual machine resizes in an Azure resource group. The `Azure.ResourceManager.Compute` namespace models Azure virtual machine as `VirtualMachineResource` and `Azure.ResourceManager` namespace models Azure resource group as `ResourceGroupResource`.  To query the virtual machines for a resource group, you would write:

Reviewer 2: In this 1st sentence, is "resizes" supposed to be "resource"?

Author: Yeah this should be a typo. Fixed

:::code language="csharp" source="snippets/unit-testing/ResourceManager/ResourceManagerCodeStructure.cs" id="ParentOfVMIsRG" :::

> Because the virtual machine-related functionality such as `GetVirtualMachines` on `ResourceGroupResource` are implemented as extension methods, it is not possible to just create a mock of the type and override the method. Instead, you'll also have to create a mock class for the "mockable resource" and wire them together.

Reviewer 2: ```suggestion
Because the virtual machine-related functionality such as `GetVirtualMachines` on `ResourceGroupResource` are implemented as extension methods, it's not possible to just create a mock of the type and override the method. Instead, you'll also have to create a mock class for the "mockable resource" and wire them together.
```

> The mockable resource type is always in the `Mocking` sub-namespace of the extension method. In the above example, the mockable resource type is in the `Azure.ResourceManager.Compute.Mocking` namespace. And the mockable resource type is always named after the resource type with a "Mockable" and the library name as prefixes. In the above example, the mockable resource type is named `MockableComputeResourceGroupResource`, where `ResourceGroupResource` is the resource type of the extension method, and `Compute` is the library name.

Reviewer 2: When possible, avoid directional terms like "above".
```suggestion
The mockable resource type is always in the `Mocking` sub-namespace of the extension method. In the preceding example, the mockable resource type is in the `Azure.ResourceManager.Compute.Mocking` namespace. The mockable resource type is always named after the resource type with "Mockable" and the library name as prefixes. In the preceding example, the mockable resource type is named `MockableComputeResourceGroupResource`, where `ResourceGroupResource` is the resource type of the extension method, and `Compute` is the library name.
```

> One more thing to do before you could get the unit test running is that you need to mock the `GetCachedClient` method on the resource type of the extension method to hook up the extension method and the method on the mockable resource type.

Reviewer 2: ```suggestion
One more requirement before you get the unit test running is to mock the `GetCachedClient` method on the resource type of the extension method. Completing this step hooks up the extension method and the method on the mockable resource type.
```

> Here is how it works:

Reviewer 2: ```suggestion
Here's how it works:
```

This approach is useful when you would like to consolidate the dependency creation and share the client between multiple consuming classes.

> ## Explore the extension methods in Azure Resource Manager SDKs

Reviewer: ```suggestion
## Understand Azure Resource Manager clients
```

We should use the same form as the "Understand service clients" here; that means, we'll need an overview.

It is also important to note that the "Azure SDK" is the all-up product name.  Individual packages are "client libraries" or "management libraries" - we've been asked not to refer to them as "SDKs".

Author: Thank you! Fixed

> For fundamental concepts of Azure Resource Manager SDKs, see [Key Concepts of Azure Resource Manager SDKs](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/resourcemanager/Azure.ResourceManager#key-concepts).

Reviewer: I don't understand how this relates to unit testing.  Our goal is not to introduce the management libraries - I think this should be removed.

Also - same note on the use of "SDKs"

Author: My intention here is to add a brief introduction of the key concepts so that people knows why we have to use extension methods.
I will remove this part.

> In Azure Resource Manager SDKs, we design the clients (resources and collections) to have their hierarchical structures to emphasize their relationship between each other. To achieve that, extension methods are widely used because the parent resource of one resource might not exist in the same SDK library.

Reviewer: ```suggestion
In Azure Resource Manager libraries, the clients were designed to emphasize their relationship to one another, mirroring the service hierarchy.  To achieve that goal, extension methods are widely used to add additional features to clients.
```

> For instance, in `Azure.ResourceManager.Compute` namespace, there is a resource `VirtualMachineResource` which stands for a virtual machine resource on Azure, and its parent is `ResourceGroupResource` in `Azure.ResourceManager` namespace, and we could get the `VirtualMachineCollection` on the `ResourceGroupResource` in this way:

Reviewer: ```suggestion
For example, an Azure virtual machine resizes in an Azure resource group.   The `Azure.ResourceManager.Compute` namespace models these types as `VirtualMachineResource` and `ResourceGroupResource`, respectively.  To query the virtual machines for a resource group, you would write:
```

Author: `ResourceGroupResource` class lives in `Azure.ResourceManager` assembly, I added this information in this line as well

:::code language="csharp" source="snippets/unit-testing/ResourceManager/ResourceManagerCodeStructure.cs" id="ParentOfVMIsRG" :::

To make sure these extension methods are able to be mocked or unit tested, these methods are always implemented in this way:

> :::code language="csharp" source="snippets/unit-testing/ResourceManager/ComputeExtensions.cs" id="HowExtensionMethodsAreImplemented" :::

Reviewer 2: Wrap longer lines of code to prevent horizontal scrolling in the doc. Here's how the code currently displays:

![image](https://github.com/dotnet/docs/assets/10702007/ad5cc43d-d1c7-4426-8f74-456b6555bdbc)

Author: does this width vary on different monitor or resolution?
On my preview, some of them need to be wrapped but not as this much.
For instance, the `public static VirtualMachineCollection GetVirtualMachines` method could fit fine on my preview.

Do we have a recommended number for how many characters we could have on one line?

BTW I wrapped some lines based on the preview on my side.

Author: And when we have to wrap a long line of a method declaration, where should we put the `(`? On the first line or the second?
I am currently doing this:
```csharp
    internal ComputeResourceGroupMockingExtension(
        ArmClient client, ResourceIdentifier id) : base(client, id)
    { }
```
should I do this?
```csharp
    internal ComputeResourceGroupMockingExtension
        (ArmClient client, ResourceIdentifier id) : base(client, id)
    { }
```

Reviewer 2: The recommended number of chars is 85. See https://review.learn.microsoft.com/help/platform/code-in-docs?branch=main#horizontal-scroll-bars. My preference is to put `(` on the first line.

Author: These details now are hidden and no longer show up on the page therefore I believe it is fine. I will pay attention to other places to see if we have horizontal scoll bars

> To make sure these extension methods are able to be mocked or unit tested, these methods are always implemented in this way:
> 
> :::code language="csharp" source="snippets/unit-testing/ResourceManager/ComputeExtensions.cs" id="HowExtensionMethodsAreImplemented" :::
> 
> All the extension methods in Azure Resource Manager SDKs are implemented in the same pattern:
> 
> 1. There is a class holding the actual implementation of the method, we call it the "Mocking Extension" class. It is named in this pattern "{RP Name}{Resource Name}MockingExtension". In the above example, it is `ComputeResourceGroupMockingExtension`.
> 1. The extension method has a private method constructing an instance of the "Mocking Extension" class by calling its `GetCachedClient` method.

Reviewer: ```suggestion
```

I'm not sure that I understand why I need to know how they're implemented to understand how to test them.   I think this section and example should be removed.

Author: The details about the internal implementations has been removed.

> 1. The extension method has a private method constructing an instance of the "Mocking Extension" class by calling its `GetCachedClient` method.

Author: These have been rewritten

> Both of the methods above are instance methods, therefore they are mockable and unit-testable. In this extension methods scenario, to create a test service client, you need to create two of them, one for the "Mocking Extension", and the other is the client that this extension method is extending.

Author: these have been removed

## [Non-library](#tab/csharp)

:::code language="csharp" source="snippets/unit-testing/ResourceManager/NonLibrary/MockComputeResourceGroupMockingExtension.cs" :::

## [Moq](#tab/moq)

> :::code language="csharp" source="snippets/unit-testing/ResourceManager/Moq/MockComputeResourceGroupMockingExtension_TestSnippets_Moq.cs" id="Moq_GetVirtualMachines" :::

Reviewer 2: The `using` statements are displayed for the non-library approach. We should also display `using` statements for Moq and NSubstitute code samples.

Author: I would like to change, but how could I do that?

The usings and the code we would like to show stay at different levels on the same file - how could we put them together in the same snippet window?

Also in the `Understand service clients` section, we having usings in the `NonLibrary` part and no usings in `Moq` or `NSubstitute`. Should we change that part as well?

Reviewer 2: > The usings and the code we would like to show stay at different levels on the same file - how could we put them together in the same snippet window?

In this example, remove the `id="Moq_GetVirtualMachines"` attribute from your `:::code` markup. Then you can remove the `// <Moq_GetVirtualMachines>` snippet markers from the code file. The contents of the entire file will display, which is the desired outcome.

# [NSubstitute](#tab/nsubstitute)

:::code language="csharp" source="snippets/unit-testing/ResourceManager/NSubstitute/MockComputeResourceGroupMockingExtension_TestSnippets_NSubstitute.cs" id="NSubstitute_GetVirtualMachines" :::

---

> There is a universal naming pattern across all the Azure Resource Manager SDKs for the "mocking extension" type:

Author: These contents have been rewritten

> 1. Find the extended type of the extension method, in the above example, it is `ResourceGroupResource`.

Author: I kind of removed these steps because it is repeating the things we have stated above.

2. Trim the `Resource` suffix off it and get the "Resource Name", in the above example, it becomes `ResourceGroup`.

> 3. Find the namespace of the extension method, in the above example, it is `Azure.ResourceManager.Compute`.

Author: These have been rewritten

4. Remove the `Azure.ResourceManager` from the namespace above, and concatenate the left segments together to get the "RP Name", in the above example, it is `Compute`.
5. Now we get the name of the mocking extension type using this pattern: `{RP Name}{Resource Name}MockingExtension`. In the above extension, the name of the mocking extension type is: `ComputeResourceGroupMockingExtension`.

> 6. The mocking extension type is always in the `Mocking` sub-namespace of the extension method. In the above example, the mocking extension type is in the `Azure.ResourceManager.Compute.Mocking` namespace.

Author: well unfortunately yes, for our current solution of enabling the mocking for extension method, our customers have to know some of the internal implementation, which is also the reason I included a part of "how the extension methods are implemented".

I will try to break these steps into pieces while one of them is explaining the name pattern of the "mocking extension" and the other explaining the full process of switching the behavior of an extension method

> 6. The mocking extension type is always in the `Mocking` sub-namespace of the extension method. In the above example, the mocking extension type is in the `Azure.ResourceManager.Compute.Mocking` namespace.

Author: Sorry I am kind of occupied by something else
Here are some updates:
we are planning to write useful information on the xml document of the method itself so that our customer no longer need to know a name pattern here to get the type of mocking extension - they will find it on the xml doc of the extension method.
